File: app/app/src/main/java/com/example/hakonsreader/api/requestmodels/SubredditRequest.java
package com.example.hakonsreader.api.requestmodels;

import com.example.hakonsreader.api.RedditApi;
import com.example.hakonsreader.api.enums.PostTimeSort;
import com.example.hakonsreader.api.exceptions.InvalidAccessTokenException;
import com.example.hakonsreader.api.exceptions.NoSubredditInfoException;
import com.example.hakonsreader.api.exceptions.SubredditNotFoundException;
import com.example.hakonsreader.api.interfaces.OnFailure;
import com.example.hakonsreader.api.interfaces.OnResponse;
import com.example.hakonsreader.api.model.AccessToken;
import com.example.hakonsreader.api.model.RedditListing;
import com.example.hakonsreader.api.model.RedditPost;
import com.example.hakonsreader.api.model.Subreddit;
import com.example.hakonsreader.api.responses.GenericError;
import com.example.hakonsreader.api.responses.ListingResponse;
import com.example.hakonsreader.api.service.ImgurService;
import com.example.hakonsreader.api.service.SubredditService;
import com.example.hakonsreader.api.utils.Util;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.internal.EverythingIsNonNull;

/**
 * Class that provides an interface towards the Reddit API related to subreddits, such as retrieving
 * posts for the subreddit and subscribing to the subreddit
 */
public class SubredditRequest {

    private final AccessToken accessToken;
    private final SubredditService api;
    private final ImgurRequest imgurRequest;
    private final String subredditName;
    private final boolean loadImgurAlbumsAsRedditGalleries;

    /**
     *
     * @param accessToken The access token to use for requests
     * @param api The API service to use for requests
     * @param subredditName The name of the subreddit to make requests towards
     * @param imgurService The service to optionally use for loading Imgur albums directly. Set to
     *                     {@code null} to not load albums.
     */
    public SubredditRequest(AccessToken accessToken, SubredditService api, String subredditName, ImgurService imgurService) {
        this.accessToken = accessToken;
        this.api = api;
        this.subredditName = subredditName;
        this.imgurRequest = new ImgurRequest(imgurService);
        this.loadImgurAlbumsAsRedditGalleries = imgurService != null;
    }

    /**
     * Retrieve information about the subreddit
     *
     * <p>OAuth scope required: {@code read}</p>
     *
     * @param onResponse The response handler for successful requests. Holds the {@link Subreddit} retrieved
     * @param onFailure The response handler for failed requests. If the function is called with a "standard"
     *                  subreddit (front page, popular, all) this will be called
     */
    public void info(OnResponse<Subreddit> onResponse, OnFailure onFailure) {
        // Ensure the subreddit name is lowercased when matching against standard subs
        if (RedditApi.STANDARD_SUBS.contains(subredditName.toLowerCase())) {
            onFailure.onFailure(new GenericError(-1), new NoSubredditInfoException("The subreddits: " + RedditApi.STANDARD_SUBS.toString() + " do not have any info to retrieve"));
            return;
        }

        api.getSubredditInfo(subredditName).enqueue(new Callback<RedditListing>() {
            @Override
            public void onResponse(Call<RedditListing> call, Response<RedditListing> response) {
                RedditListing body = null;
                if (response.isSuccessful()) {
                    body = response.body();
                }

                if (body != null) {
                    if (body.getId() != null) {
                        onResponse.onResponse((Subreddit) body);
                    } else {
                        onFailure.onFailure(new GenericError(response.code()), new SubredditNotFoundException("No subreddit found with name: " + subredditName));
                    }
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<RedditListing> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
    }

    /**
     * NOTE: the response for this request is sent on a background thread
     *
     * <p>Retrieves posts from the subreddit. The posts here are sorted by "hot", if you want to retrieve posts with a
     * different sort use {@link SubredditRequest#posts()}</p>
     *
     * <p>If an access token is set posts are customized for the user</p>
     *
     * <p>No specific OAuth scope is required</p>
     *
     * @param after The ID of the last post seen (or an empty string if first time loading)
     * @param count The amount of posts already retrieved (0 if first time loading)
     * @param onResponse The callback for successful requests. Holds a {@link List} of {@link RedditPost} objects
     * @param onFailure The callback for failed requests
     */
    @EverythingIsNonNull
    public void posts(String after, int count, OnResponse<List<RedditPost>> onResponse, OnFailure onFailure) {
        this.getPosts("hot", null, after, count, onResponse, onFailure);
    }

    /**
     * <p>NOTE: the response for request is sent on a background thread</p>
     *
     * Retrieve an object to make API calls for posts in the subreddit
     *
     * @return An object that can retrieve new, top, and controversial posts for the subreddit
     */
    public SubredditPostsRequets posts() {
        return new SubredditPostsRequets();
    }

    /**
     * Favorite or un-favorite a subreddit
     *
     * @param favorite True if the action should be to favorite, false to un-favorite
     * @param onResponse The response handler for successful requests. Does not hold any data, but will
     *                   be called when the request succeeds.
     * @param onFailure Callback for failed requests
     */
    public void favorite(boolean favorite, OnResponse<Void> onResponse, OnFailure onFailure) {
        try {
            Util.verifyLoggedInToken(accessToken);
        } catch (InvalidAccessTokenException e) {
            onFailure.onFailure(new GenericError(-1), new InvalidAccessTokenException("Favoriting a subreddit requires a valid access token for a logged in user", e));
            return;
        }

        api.favoriteSubreddit(subredditName, favorite).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if (response.isSuccessful()) {
                    onResponse.onResponse(null);
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
    }

    /**
     * Subscribe or unsubscribe to a subreddit
     *
     * <p>OAuth scope required: {@code subscribe}</p>
     *
     * @param subscribe True if the action should be to subscribe, false to unsubscribe
     * @param onResponse The response handler for successful requests. Does not hold any data, but will
     *                   be called when the request succeeds.
     * @param onFailure Callback for failed requests
     */
    public void subscribe(boolean subscribe, OnResponse<Void> onResponse, OnFailure onFailure) {
        try {
            Util.verifyLoggedInToken(accessToken);
        } catch (InvalidAccessTokenException e) {
            onFailure.onFailure(new GenericError(-1), new InvalidAccessTokenException("Subscribing to a subreddit requires a valid access token for a logged in user", e));
            return;
        }

        api.subscribeToSubreddit(subscribe ? "sub" : "unsub", subredditName).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if (response.isSuccessful()) {
                    onResponse.onResponse(null);
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
    }


    /**
     * <p>NOTE: the response for this request is sent on a background thread</p>
     *
     * Retrieves posts from the subreddit
     *
     * <p>If an access token is set posts are customized for the user</p>
     *
     * <p>No specific OAuth scope is required</p>
     *
     * @param sort The sort for the posts (new, hot, top, or controversial)
     * @param timeSort How the posts should be time sorted. This only has an affect on top and controversial,
     *                 and can be set to null for new and hot
     * @param after The ID of the last post seen (or an empty string if first time loading)
     * @param count The amount of posts already retrieved (0 if first time loading)
     * @param onResponse The callback for successful requests. Holds a {@link List} of {@link RedditPost} objects
     * @param onFailure The callback for failed requests. If the subreddit doesn't exist this will be called
     */
    private void getPosts(String sort, PostTimeSort timeSort, String after, int count, OnResponse<List<RedditPost>> onResponse, OnFailure onFailure) {
        // Not front page, add r/ prefix
        String subreddit = subredditName;

        if (!subreddit.isEmpty()) {
            subreddit = "r/" + subreddit;
        }

        String finalSubreddit = subreddit;

        // Loading Imgur albums requires API calls inside the callback. If we use "enqueue" and operate
        // on the current thread the RedditPost objects will be updated after the response is given with
        // onResponse, which means the UI potentially wont be correct, so we have to run this entire thing on
        // a background thread
        new Thread(() -> {
            try {
                Response<ListingResponse> response = api.getPosts(
                        finalSubreddit,
                        sort,
                        timeSort == null ? "" : timeSort.getValue(),
                        after,
                        count,
                        RedditApi.RAW_JSON
                ).execute();

                okhttp3.Response prior = response.raw().priorResponse();

                // If the subreddit doesn't exist, Reddit wants to be helpful (or something) and redirects
                // the response to a search request instead. This causes issues as the response being sent back
                // now holds subreddits instead of posts, so if we have a prior request (which is the actual original request)
                // then call the failure handler as the user of the API might want to know that the sub doesn't exist
                // Additionaly, if the search request returned subreddits, body.getListings() will hold a List<Subreddit> which will cause issues
                // This is also an "issue" for SubredditRequest.info(), but it will manage to convert that to a RedditListing
                // and the check for getId() will return null, so it doesn't have to be handled directly
                // We could disable redirects, but I'm afraid of what issues that would cause later
                if (prior != null) {
                    // TODO when new access tokens are retrieved automatically, the prior response is also set which means
                    //  the posts aren't returned
                    onFailure.onFailure(new GenericError(response.code()), new SubredditNotFoundException("No subreddit found with name: " + subredditName));
                    return;
                }

                ListingResponse body = null;
                if (response.isSuccessful()) {
                    body = response.body();
                }

                if (body != null) {
                    List<RedditPost> posts = (List<RedditPost>) body.getListings();

                    if (loadImgurAlbumsAsRedditGalleries) {
                        imgurRequest.loadAlbums(posts);
                    }

                    onResponse.onResponse(posts);
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            } catch (IOException e) {
                onFailure.onFailure(new GenericError(-1), e);
            }
        }).start();
    }


    /**
     * Class to retrieve posts from the subreddit. The functions declared here define how to sort the posts
     * (new, hot etc.)
     */
    public class SubredditPostsRequets {

        /**
         * NOTE: the response for this request is sent on a background thread
         *
         * <p>Get the "controversial" posts for the user</p>
         *
         * <p>OAuth scope required: {@code history}</p>
         *
         * @param after The ID of the last post seen (or an empty string if first time loading)
         * @param timeSort The time sort for the posts (of all time, of hour etc.)
         * @param count The amount of posts already retrieved (0 if first time loading)
         * @param onResponse The callback for successful requests. Holds a {@link List} of {@link RedditPost} objects
         * @param onFailure The callback for failed requests
         */
        public void controversial(PostTimeSort timeSort, String after, int count, OnResponse<List<RedditPost>> onResponse, OnFailure onFailure) {
            getPosts("controversial", timeSort, after, count, onResponse, onFailure);
        }

        /**
         * NOTE: the response for this request is sent on a background thread
         *
         * <p>Retrieves new posts from the subreddit</p>
         *
         * <p>If an access token is set posts are customized for the user</p>
         *
         * <p>No specific OAuth scope is required</p>
         *
         * @param after The ID of the last post seen (or an empty string if first time loading)
         * @param count The amount of posts already retrieved (0 if first time loading)
         * @param onResponse The callback for successful requests. Holds a {@link List} of {@link RedditPost} objects
         * @param onFailure The callback for failed requests
         */
        @EverythingIsNonNull
        public void newPosts(String after, int count, OnResponse<List<RedditPost>> onResponse, OnFailure onFailure) {
            getPosts("new", null, after, count, onResponse, onFailure);
        }

        /**
         * NOTE: the response for this request is sent on a background thread
         *
         * <p>Retrieves top posts from the subreddit</p>
         *
         * <p>If an access token is set posts are customized for the user</p>
         *
         * <p>No specific OAuth scope is required</p>
         *
         * @param after The ID of the last post seen (or an empty string if first time loading)
         * @param timeSort The time sort for the posts (of all time, of hour etc.)
         * @param count The amount of posts already retrieved (0 if first time loading)
         * @param onResponse The callback for successful requests. Holds a {@link List} of {@link RedditPost} objects
         * @param onFailure The callback for failed requests
         */
        @EverythingIsNonNull
        public void top(PostTimeSort timeSort, String after, int count, OnResponse<List<RedditPost>> onResponse, OnFailure onFailure) {
            getPosts("top", timeSort, after, count, onResponse, onFailure);
        }
    }
}

File: app/app/src/main/java/com/example/hakonsreader/api/requestmodels/SubredditRequest.kt
package com.example.hakonsreader.api.requestmodels

import com.example.hakonsreader.api.RedditApi
import com.example.hakonsreader.api.enums.FlairType
import com.example.hakonsreader.api.enums.PostTimeSort
import com.example.hakonsreader.api.enums.SortingMethods
import com.example.hakonsreader.api.enums.Thing
import com.example.hakonsreader.api.exceptions.InvalidAccessTokenException
import com.example.hakonsreader.api.exceptions.NoSubredditInfoException
import com.example.hakonsreader.api.exceptions.SubredditNotFoundException
import com.example.hakonsreader.api.model.*
import com.example.hakonsreader.api.model.flairs.RedditFlair
import com.example.hakonsreader.api.model.thirdparty.ThirdPartyOptions
import com.example.hakonsreader.api.requestmodels.thirdparty.ThirdPartyRequest
import com.example.hakonsreader.api.responses.ApiResponse
import com.example.hakonsreader.api.responses.GenericError
import com.example.hakonsreader.api.service.thirdparty.GfycatService
import com.example.hakonsreader.api.service.SubredditService
import com.example.hakonsreader.api.service.thirdparty.ImgurService
import com.example.hakonsreader.api.utils.apiError
import com.example.hakonsreader.api.utils.apiListingErrors
import com.example.hakonsreader.api.utils.createFullName
import com.example.hakonsreader.api.utils.verifyLoggedInToken


/**
 * Interface for communicating with a subreddit
 */
interface SubredditRequest {

    /**
     * Retrieve information about the subreddit
     *
     * OAuth scope required: *read*
     */
    suspend fun info() : ApiResponse<Subreddit>

    /**
     * Retrieve subreddit rules
     *
     * OAuth scope required: *read*
     */
    suspend fun rules() : ApiResponse<List<SubredditRule>>

    /**
     * Retrieves posts from the subreddit
     *
     * If an access token for a user is set posts are customized for the user
     *
     * OAuth scope required: *read*
     *
     * @param postSort sort for the posts (new, hot, top, or controversial). Default is [SortingMethods.HOT]
     * @param timeSort How the posts should be time sorted. This only has an affect on top and controversial.
     * Default is [PostTimeSort.DAY]
     * @param after The ID of the last post seen. Default is an empty string (ie. no last post)
     * @param count The amount of posts already retrieved. Default is *0* (ie. no posts already)
     * @param limit The amount of posts to retrieve. Default is *25*
     */
    suspend fun posts(postSort: SortingMethods = SortingMethods.HOT, timeSort: PostTimeSort = PostTimeSort.DAY, after: String = "", count: Int = 0, limit: Int = 25) : ApiResponse<List<RedditPost>>

    /**
     * Subscribe or unsubscribe to the subreddit
     *
     * OAuth scope required: *subscribe*
     *
     * @param subscribe True to subscribe, false to unsubscribe
     * @return If successful no value is returned
     */
    suspend fun subscribe(subscribe: Boolean) : ApiResponse<Unit>

    /**
     * Favorite or un-favorite a subreddit
     *
     * @param favorite True if the action should be to favorite, false to un-favorite
     * @return If successful no value is returned
     */
    suspend fun favorite(favorite: Boolean) : ApiResponse<Unit>

    /**
     * Submit a text post to the subreddit
     *
     * OAuth scope required: *submit*
     *
     * @param title The title of the post. Max characters is 300
     * @param text The text of the post. Can be omitted for title-only posts
     * @param nsfw True if the post should be marked as NSFW (18+). Default to *false*
     * @param spoiler True if the post should be marked as a spoiler. Default to *false*
     * @param receiveNotifications True if the user wants to receive notifications from the post. Default to *true*
     * @param flairId The ID of the flair to submit as the link flair for the post. Should be the the value
     * retrieved with [RedditFlair.id]. Default is an empty string (ie. no flair)
     *
     * @return An [ApiResponse] holding a [Submission] of the newly created post
     */
    suspend fun submitTextPost(
            title: String,
            text: String = "",

            nsfw: Boolean = false,
            spoiler: Boolean = false,
            receiveNotifications: Boolean = true,

            flairId: String = ""
    ) : ApiResponse<Submission>

    /**
     * Submit a link post to the subreddit
     *
     * OAuth scope required: *submit*
     *
     * @param title The title of the post. Max characters is 300
     * @param link The link the post is referencing. This should be a valid URL, verification is not done
     * by the API. Spaces in the link will be converted to *%20*
     * @param nsfw True if the post should be marked as NSFW (18+). Default to *false*
     * @param spoiler True if the post should be marked as a spoiler. Default to *false*
     * @param receiveNotifications True if the user wants to receive notifications from the post. Default to *true*
     * @param flairId The ID of the flair to submit as the link flair for the post. Should be the the value
     * retrieved with [RedditFlair.id]. Default is an empty string (ie. no flair)
     *
     * @return An [ApiResponse] holding a [Submission] of the newly created post
     */
    suspend fun submitLinkPost(
            title: String,
            link: String,

            nsfw: Boolean = false,
            spoiler: Boolean = false,
            receiveNotifications: Boolean = true,

            flairId: String = ""
    ) : ApiResponse<Submission>

    /**
     * Submit a text post to the subreddit
     *
     * OAuth scope required: *submit*
     *
     * @param title The title of the post. Max characters is 300
     * @param crosspostId The ID of the post this post is crossposting
     * @param nsfw True if the post should be marked as NSFW (18+). Default to *false*
     * @param spoiler True if the post should be marked as a spoiler. Default to *false*
     * @param receiveNotifications True if the user wants to receive notifications from the post. Default to *true*
     * @param flairId The ID of the flair to submit as the link flair for the post. Should be the the value
     * retrieved with [RedditFlair.id]. Default is an empty string (ie. no flair)
     *
     * @return An [ApiResponse] holding a [Submission] of the newly created post
     */
    suspend fun submitCrosspost(
            title: String,
            crosspostId: String,

            nsfw: Boolean = false,
            spoiler: Boolean = false,
            receiveNotifications: Boolean = true,

            flairId: String = ""
    ) : ApiResponse<Submission>

    /**
     * Gets the submission flairs for the subreddit
     *
     * OAuth scope required: *flair*
     *
     * If the subreddit doesn't allow submission flairs, a 403 Forbidden error is returned
     */
    suspend fun submissionFlairs() : ApiResponse<List<RedditFlair>>

    /**
     * Gets the user flairs for the subreddit
     *
     * OAuth scope required: *flair*
     *
     * If the subreddit doesn't allow submission flairs, a 403 Forbidden error is returned
     */
    suspend fun userFlairs() : ApiResponse<List<RedditFlair>>

    /**
     * Select a flair for a user on the subreddit. If [flairId] is not called with `null`, then
     * [enableUserFlair] is called with `true` automatically
     *
     * OAuth scope required: *flair*
     *
     * @param username The username to select a flair for
     * @param flairId The ID of the flair to select (as from [RedditFlair.id]), or `null` to clear
     * the users flair
     */
    suspend fun selectFlair(username: String, flairId: String?) : ApiResponse<Unit>

    /**
     * Enables or disables user flairs on the subreddit
     *
     * @param enable True to enable flairs, false to disable
     */
    suspend fun enableUserFlair(enable: Boolean) : ApiResponse<Unit>

    /**
     * Gets a wiki page for the subreddit
     *
     * OAuth scope required: `wikiread`
     *
     * @param page The name of the page to retrieve. Default to `index` (the start of the wiki)
     */
    suspend fun wiki(page: String = "index") : ApiResponse<SubredditWikiPage>
}

/**
 * Standard [SubredditRequest] implementation
 */
class SubredditRequestImpl(
        private val subredditName: String,
        private val accessToken: AccessToken,
        private val api: SubredditService,
        imgurApi: ImgurService?,
        gfycatApi: GfycatService,
        thirdPartyOptions: ThirdPartyOptions
) : SubredditRequest {

    private val thirdPartyRequest = ThirdPartyRequest(imgurApi, gfycatApi, thirdPartyOptions)

    override suspend fun info() : ApiResponse<Subreddit> {
        if (RedditApi.STANDARD_SUBS.contains(subredditName.toLowerCase())) {
            return ApiResponse.Error(GenericError(-1), NoSubredditInfoException("The subreddits: " + RedditApi.STANDARD_SUBS.toString() + " do not have any info to retrieve"))
        }
        return try {
            val resp = api.getSubredditInfo(subredditName)
            val sub = resp.body()

            if (sub != null) {
                sub as Subreddit

                // If there is no name, the subreddit wasn't found (this happens on redirects since the subreddit wasn't found
                // and the cast is successful, but the object is empty with default values)
                if (sub.name.isNotBlank()) {
                    ApiResponse.Success(sub)
                } else {
                    ApiResponse.Error(GenericError(-1), SubredditNotFoundException("The subreddit '$subredditName' was not found"))
                }
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun rules() : ApiResponse<List<SubredditRule>>  {
        return try {
            val resp = api.getRules(subredditName)
            val rules = resp.body()?.rules

            if (rules != null) {
                // Rules aren't connected automatically to its subreddit
                rules.forEach { it.subreddit = subredditName }

                ApiResponse.Success(rules)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun posts(postSort: SortingMethods, timeSort: PostTimeSort, after: String, count: Int, limit: Int) : ApiResponse<List<RedditPost>> {
        // If not blank (ie. front page) add "r/" at the start
        val sub = if (subredditName.isBlank()) {
            ""
        } else {
            "r/$subredditName"
        }

        return try {
            val resp = api.getPosts(
                    sub,
                    postSort.value,
                    timeSort.value,
                    after,
                    count,
                    limit
            )

            // If the subreddit doesn't exist, Reddit wants to be helpful (or something) and redirects
            // the response to a search request instead. This causes issues as the response being sent back
            // now holds subreddits instead of posts, so if we have a prior request (which is the actual original request)
            // then call the failure handler as the user of the API might want to know that the sub doesn't exist
            // Additionally, if the search request returned subreddits, body.getListings() will hold a List<Subreddit> which will cause issues
            // This is also an "issue" for SubredditRequest.info(), but it will manage to convert that to a RedditListing
            // and the check for getId() will return null, so it doesn't have to be handled directly
            // We could disable redirects, but I'm afraid of what issues that would cause later (edit: god prediction tbh)

            // Some redirects are on purpose, such as going to "r/random" which will redirect to a random subreddit
            val prior = resp.raw().priorResponse()
            if (prior != null) {

                // If the prior response was a redirect then we exit. "prior" can be a 401 Unauthorized because the
                // access token was invalid and automatically refreshed, so if that happens we don't want to exit
                // as the new response will hold the posts and be correct
                val code = prior.code()
                if (code in 300..399) {
                    // If the redirect is because it doesn't exist return, otherwise we can continue
                    val pathSegments = resp.raw().request().url().pathSegments()
                    if (pathSegments.size >= 2 && pathSegments[0] == "subreddits" && pathSegments[1] == "search") {
                        return ApiResponse.Error(GenericError(code), SubredditNotFoundException("No subreddit found with name: $subredditName"))
                    }
                }
            }

            val posts = resp.body()?.getListings()

            if (posts != null) {
                thirdPartyRequest.loadAll(posts)
                ApiResponse.Success(posts)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }


    override suspend fun subscribe(subscribe: Boolean) : ApiResponse<Unit> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), e)
        }

        val subAction = if (subscribe) {
            "sub"
        } else {
            "unsub"
        }

        return try {
            val resp = api.subscribeToSubreddit(subAction, subredditName)

            if (resp.isSuccessful) {
                ApiResponse.Success(Unit)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun favorite(favorite: Boolean) : ApiResponse<Unit> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), e)
        }

        return try {
            val resp = api.favoriteSubreddit(favorite, subredditName)

            if (resp.isSuccessful) {
                ApiResponse.Success(Unit)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }


    // TODO possible submit errors:
    //  {"json": {"errors": [["BAD_SR_NAME", "det navnet kommer ikke til \u00e5 virke", "sr"]]}}

    // TODO submit response: {"json": {"errors": [], "data": {"url": "https://www.reddit.com/r/hakonschia/comments/k1yj0s/hello_reddit/", "drafts_count": 0, "id": "k1yj0s", "name": "t3_k1yj0s"}}}
    //  can probably just return the ID of the post

    override suspend fun submitTextPost(
            title: String,
            text: String,

            nsfw: Boolean,
            spoiler: Boolean,
            receiveNotifications: Boolean,

            flairId: String
    ) : ApiResponse<Submission> {
        val submissionError = verifyGenericSubmission(title)
        if (submissionError != null) {
            return submissionError
        }

        return try {
            val resp = api.submit(
                    subredditName,
                    kind = "self",
                    title = title,
                    text = text,
                    nsfw = nsfw,
                    spoiler = spoiler,
                    sendNotifications = receiveNotifications,
                    flairId = flairId
            )
            val body = resp.body()?.getListing()

            if (body != null) {
                ApiResponse.Success(body)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun submitLinkPost(
            title: String,
            link: String,

            nsfw: Boolean,
            spoiler: Boolean,
            receiveNotifications: Boolean,

            flairId: String
    ) : ApiResponse<Submission> {
        val submissionError = verifyGenericSubmission(title)
        if (submissionError != null) {
            return submissionError
        }

        val linkSpaceConverted = link.replace(" ", "%20")

        return try {
            val resp = api.submit(
                    subredditName,
                    kind = "link",
                    title = title,
                    link = linkSpaceConverted,
                    nsfw = nsfw,
                    spoiler = spoiler,
                    sendNotifications = receiveNotifications,
                    flairId = flairId
            )
            val body = resp.body()?.getListing()

            if (body != null) {
                ApiResponse.Success(body)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun submitCrosspost(
            title: String,
            crosspostId: String,

            nsfw: Boolean,
            spoiler: Boolean,
            receiveNotifications: Boolean,

            flairId: String
    ) : ApiResponse<Submission> {
        val submissionError = verifyGenericSubmission(title)
        if (submissionError != null) {
            return submissionError
        }

        // kind = crosspost
        // "crosspost_fullname"
        val fullname = createFullName(Thing.POST, crosspostId)

        return try {
            val resp = api.submit(
                    subredditName,
                    kind = "crosspost",
                    title = title,
                    crosspostFullname = fullname,
                    nsfw = nsfw,
                    spoiler = spoiler,
                    sendNotifications = receiveNotifications,
                    flairId = flairId
            )
            val body = resp.body()?.getListing()

            if (body != null) {
                ApiResponse.Success(body)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }

        // TODO possible error:
        //  {"json": {"errors": [["INVALID_CROSSPOST_THING", "that isn't a valid crosspost url", "crosspost_thing"]]}}
    }

    private fun verifyGenericSubmission(title: String) : ApiResponse.Error? {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Submitting a post requires a valid access token for a logged in user", e))
        }

        if (title.length > 300) {
            return ApiResponse.Error(GenericError(-1), IllegalStateException("Post titles cannot be longer than 300 characters"))
        } else if (title.isBlank()) {
            return ApiResponse.Error(GenericError(-1), IllegalStateException("Post titles cannot be empty"))
        }

        return null
    }


    override suspend fun submissionFlairs() : ApiResponse<List<RedditFlair>> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Retrieving submission flairs requires a valid access token for a logged in user", e))
        }

        return try {
            val response = api.getLinkFlairs(subredditName)
            val flairs = response.body()

            if (flairs != null) {
                flairs.forEach {
                    it.subreddit = subredditName
                    it.flairType = FlairType.SUBMISSION
                }
                ApiResponse.Success(flairs)
            } else {
                apiError(response)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun userFlairs() : ApiResponse<List<RedditFlair>> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Retrieving user flairs requires a valid access token for a logged in user", e))
        }

        return try {
            val response = api.getUserFlairs(subredditName)
            val flairs = response.body()

            if (flairs != null) {
                flairs.forEach {
                    it.subreddit = subredditName
                    it.flairType = FlairType.USER
                }
                ApiResponse.Success(flairs)
            } else {
                apiError(response)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun selectFlair(username: String, flairId: String?) : ApiResponse<Unit> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Setting a flair requires a valid access token for a logged in user", e))
        }

        return try {
            val response = api.selectFlair(subredditName, username, flairId)
            val body = response.body()

            if (body != null) {
                if (!body.hasErrors()) {
                    if (flairId != null) {
                        enableUserFlair(true)
                    }
                    ApiResponse.Success(Unit)
                } else {
                    apiListingErrors(body.errors() as List<List<String>>)
                }
            } else {
                apiError(response)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun enableUserFlair(enable: Boolean) : ApiResponse<Unit> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Enabling flairs requires a valid access token for a logged in user", e))
        }

        return try {
            val response = api.enableUserFlair(subredditName, enable)
            val body = response.body()

            if (body != null) {
                if (!body.hasErrors()) {
                    ApiResponse.Success(Unit)
                } else {
                    apiListingErrors(body.errors() as List<List<String>>)
                }
            } else {
                apiError(response)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    override suspend fun wiki(page: String) : ApiResponse<SubredditWikiPage> {
        return try {
            val response = api.getWikiPage(subredditName, page)
            val body = response.body()?.data
            if (body != null) {
                body.subreddit = subredditName
                ApiResponse.Success(body)
            } else {
                apiError(response)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }
}